lib/inky/inky.ex

```diff
@@ -0,0 +1,625 @@
+  alias Inky.InkyWhat
+  alias Inky.State
+  # alias Inky.Pixel
+
+  @reset_pin 27
+  @busy_pin 17
+  @dc_pin 22
+
+  # @mosi_pin 10
+  # @sclk_pin 11
+  @cs0_pin 0
+
+  @spi_chunk_size 4096
+  @spi_command 0
+  @spi_data 1
+
+  # Used in logo example
+  # inkyphat and inkywhat classes
+  # color constants: RED, BLACK, WHITE
+  # dimension constants: WIDTH, HEIGHT
+  # PIL: putpixel(value)
+  # set_image
+  # show
+
+  # SPI bus options include:
+  # * `mode`: This specifies the clock polarity and phase to use. (0)
+  # * `bits_per_word`: bits per word on the bus (8)
+  # * `speed_hz`: bus speed (1000000)
+  # * `delay_us`: delay between transaction (10)
+
+  def setup(state \\ nil, type, luts_color)
+      when type in [:phat, :what] and luts_color in [:black, :red, :yellow] do
+    state =
+      case state do
+        %State{} ->
+          state
+
+        nil ->
+          {:ok, dc_pid} = GPIO.open(@dc_pin, :output)
+          {:ok, reset_pid} = GPIO.open(@reset_pin, :output)
+          {:ok, busy_pid} = GPIO.open(@busy_pin, :input)
+          # GPIO.write(gpio_pid, 1)
+          {:ok, spi_pid} = SPI.open("spidev0." <> to_string(@cs0_pin), speed_hz: 488_000)
+          # Use binary pattern matching to pull out the ADC counts (low 10 bits)
+          # <<_::size(6), counts::size(10)>> = SPI.transfer(spi_pid, <<0x78, 0x00>>)
+          %State{
+            dc_pid: dc_pid,
+            reset_pid: reset_pid,
+            busy_pid: busy_pid,
+            spi_pid: spi_pid,
+            color: luts_color
+          }
+      end
+
+    GPIO.write(state.reset_pid, 0)
+    :timer.sleep(100)
+    GPIO.write(state.reset_pid, 1)
+    :timer.sleep(100)
+
+    state =
+      case type do
+        :phat -> InkyPhat.update_state(state)
+        :what -> InkyWhat.update_state(state)
+      end
+
+    soft_reset(state)
+    busy_wait(state)
+    state
+  end
+
+  def set_pixel(state = %State{}, x, y, value) do
+    state =
+      if value in [state.white, state.black, state.red, state.yellow] do
+        put_in(state.pixels[{x, y}], value)
+        # %{state | pixels: %{state.pixels | {x, y}: value} }
+      else
+        state
+      end
+
+    state
+  end
+
+  def show(state = %State{}) do
+    # Not implemented: vertical flip
+    # Not implemented: horizontal flip
+
+    # Note: Rotation handled when converting to bytestring
+
+    black_bytes = pixels_to_bytestring(state, state.black, 0, 1)
+    red_bytes = pixels_to_bytestring(state, state.red, 1, 0)
+    update(state, black_bytes, red_bytes)
+  end
+
+  def log_grid(state = %State{}) do
+    grid =
+      Enum.reduce(0..(state.height - 1), "", fn y, grid ->
+        row =
+          Enum.reduce(0..(state.width - 1), "", fn x, row ->
+            color_value = Map.get(state.pixels, {x, y}, 0)
+
+            row <>
+              case color_value do
+                0 -> "W"
+                1 -> "B"
+                2 -> "R"
+              end
+          end)
+
+        grid <> row <> "\n"
+      end)
```

Comment:
I wonder if it makes sense to offer this nested `Enum.reduce` as a helper function.

Reply:
It might, yeah. Tends to be useful. I haven't really started on some of the stuff the original library does which is pushing images into the pixel grid and stuff. But a few helpers to make fills and maybe some shapes would be a reasonable next step.

Reply:
Thinking about how to design this and I'm tripping myself up on how it should work. Because we probably want to perform one operation per pixel and an additional operation per row of pixels. And the row reducer function would need to know how to the final product should be merged together. If you have a suggested approach I'd love to see it. I'm sure there are a few clever ways to manage this but I'm drawing a blank.